File: Golang/04. Association Relationship/practice-13.go
/**
 * Practice 13: Many-Many Association — Patient & Medications
 * Task: Patients prescribed multiple medications; each medication
 *       prescribed to many patients. Certain combinations forbidden.
 *
 *       Forbidden combinations:
 *         - Antibiotics & Statins
 *         - Muscle Relaxants & CNS Depressants
 *         - Anti-Inflammatories & Anticoagulants
 *
 * How to run:
 *   go run practice-13.go
 *
 * Key Concepts:
 *   - Many-Many Association via a Prescription junction struct
 *   - Conflict detection logic
 *   - Slice and map management
 */

package main

import (
	"fmt"
	"time"
)

/** Medication with name and category. */
type Medication13 struct {
	Name     string
	Category string
}

func (m *Medication13) ShowInfo() {
	fmt.Printf("  %-25s | Category: %s\n", m.Name, m.Category)
}

/** Patient who can be prescribed multiple medications. */
type Patient13 struct {
	PatientId string
	Name      string
}

/** Prescription links a patient to a medication. */
type Prescription13 struct {
	Patient        *Patient13
	Medication     *Medication13
	PrescribedDate time.Time
}

func (p *Prescription13) ShowInfo() {
	fmt.Printf("  %-12s | %-25s | Category: %-22s | Date: %s\n",
		p.Patient.Name, p.Medication.Name, p.Medication.Category,
		p.PrescribedDate.Format("2006-01-02"))
}

/** Forbidden medication category combinations. */
var forbiddenCombinations = [][2]string{
	{"Antibiotics", "Statins"},
	{"Muscle Relaxants", "CNS Depressants"},
	{"Anti-Inflammatories", "Anticoagulants"},
}

/** PrescriptionManager manages all prescriptions with conflict checking. */
type PrescriptionManager13 struct {
	Prescriptions []*Prescription13
}

func NewPrescriptionManager13() *PrescriptionManager13 {
	return &PrescriptionManager13{}
}

func (pm *PrescriptionManager13) getPatientCategories(patient *Patient13) map[string]bool {
	categories := make(map[string]bool)
	for _, p := range pm.Prescriptions {
		if p.Patient.PatientId == patient.PatientId {
			categories[p.Medication.Category] = true
		}
	}
	return categories
}

func (pm *PrescriptionManager13) checkConflict(existing map[string]bool, newCategory string) string {
	for _, pair := range forbiddenCombinations {
		if newCategory == pair[0] && existing[pair[1]] {
			return fmt.Sprintf("%s cannot be combined with %s", pair[0], pair[1])
		}
		if newCategory == pair[1] && existing[pair[0]] {
			return fmt.Sprintf("%s cannot be combined with %s", pair[1], pair[0])
		}
	}
	return ""
}

func (pm *PrescriptionManager13) Prescribe(patient *Patient13, medication *Medication13, date time.Time) {
	// Check duplicate
	for _, p := range pm.Prescriptions {
		if p.Patient.PatientId == patient.PatientId && p.Medication.Name == medication.Name {
			fmt.Printf("  [Error] %s is already prescribed to %s.\n", medication.Name, patient.Name)
			return
		}
	}
	// Check conflict
	existing := pm.getPatientCategories(patient)
	conflict := pm.checkConflict(existing, medication.Category)
	if conflict != "" {
		fmt.Printf("  [Error] Cannot prescribe %s to %s: %s.\n", medication.Name, patient.Name, conflict)
		return
	}
	pm.Prescriptions = append(pm.Prescriptions, &Prescription13{patient, medication, date})
	fmt.Printf("  [OK] %s prescribed to %s.\n", medication.Name, patient.Name)
}

func (pm *PrescriptionManager13) GetPatientPrescriptions(patient *Patient13) []*Prescription13 {
	var result []*Prescription13
	for _, p := range pm.Prescriptions {
		if p.Patient.PatientId == patient.PatientId {
			result = append(result, p)
		}
	}
	return result
}

func (pm *PrescriptionManager13) GetMedicationPatients(medication *Medication13) []*Prescription13 {
	var result []*Prescription13
	for _, p := range pm.Prescriptions {
		if p.Medication.Name == medication.Name {
			result = append(result, p)
		}
	}
	return result
}

func (pm *PrescriptionManager13) ShowAll() {
	if len(pm.Prescriptions) == 0 {
		fmt.Println("  No prescriptions recorded.")
		return
	}
	fmt.Printf("  %-12s | %-25s | %-22s | Date\n", "Patient", "Medication", "Category")
	fmt.Println("  --------------------------------------------------------------------------------")
	for _, p := range pm.Prescriptions {
		p.ShowInfo()
	}
	fmt.Println()
}

func main() {
	amoxicillin := &Medication13{"Amoxicillin", "Antibiotics"}
	ibuprofen := &Medication13{"Ibuprofen", "Anti-Inflammatories"}
	diazepam := &Medication13{"Diazepam", "Muscle Relaxants"}
	coughSyrup := &Medication13{"Dextromethorphan", "Cough Syrup"}
	insulin := &Medication13{"Insulin", "Insulin"}
	warfarin := &Medication13{"Warfarin", "Anticoagulants"}
	phenobarbital := &Medication13{"Phenobarbital", "CNS Depressants"}
	atorvastatin := &Medication13{"Atorvastatin", "Statins"}

	fmt.Println("=== Available Medications ===")
	for _, m := range []*Medication13{amoxicillin, ibuprofen, diazepam, coughSyrup,
		insulin, warfarin, phenobarbital, atorvastatin} {
		m.ShowInfo()
	}
	fmt.Println()

	patient1 := &Patient13{"PAT-001", "Tareq"}
	patient2 := &Patient13{"PAT-002", "Afsana"}
	patient3 := &Patient13{"PAT-003", "Imtiaz"}

	manager := NewPrescriptionManager13()
	today := time.Now()

	fmt.Println("=== Prescribing Medications ===")
	manager.Prescribe(patient1, amoxicillin, today)
	manager.Prescribe(patient1, coughSyrup, today)
	manager.Prescribe(patient1, atorvastatin, today) // CONFLICT
	fmt.Println()

	manager.Prescribe(patient2, diazepam, today)
	manager.Prescribe(patient2, insulin, today)
	manager.Prescribe(patient2, phenobarbital, today) // CONFLICT
	fmt.Println()

	manager.Prescribe(patient3, ibuprofen, today)
	manager.Prescribe(patient3, coughSyrup, today)
	manager.Prescribe(patient3, warfarin, today) // CONFLICT
	fmt.Println()

	fmt.Println("=== All Prescriptions ===")
	manager.ShowAll()

	for _, patient := range []*Patient13{patient1, patient2, patient3} {
		fmt.Printf("=== Prescriptions for %s ===\n", patient.Name)
		for _, p := range manager.GetPatientPrescriptions(patient) {
			p.ShowInfo()
		}
		fmt.Println()
	}

	fmt.Println("=== Patients on Dextromethorphan (Cough Syrup) ===")
	for _, p := range manager.GetMedicationPatients(coughSyrup) {
		fmt.Printf("  %s (prescribed: %s)\n", p.Patient.Name, p.PrescribedDate.Format("2006-01-02"))
	}
	fmt.Println()
}

File: Java/04. Association Relationship/Practice13.java
/**
 * Practice 13: Many-Many Association — Patient & Medications
 * Task: Patients are prescribed multiple medications, and each medication
 *       can be prescribed to many patients. Certain combinations are forbidden.
 *
 *       Forbidden combinations:
 *         - Antibiotics & Statins
 *         - Muscle Relaxants & CNS Depressants
 *         - Anti-Inflammatories & Anticoagulants
 *
 * How to compile and run:
 *   javac Practice13.java
 *   java Practice13
 *
 * Key Concepts:
 *   - Many-Many Association via a Prescription junction class
 *   - Conflict detection logic
 *   - Collection management
 */

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Practice13 {

    /** Medication with name and category. */
    static class Medication {
        private String name;
        private String category;

        public Medication(String name, String category) {
            this.name = name;
            this.category = category;
        }

        public String getName() { return name; }
        public String getCategory() { return category; }

        public void showInfo() {
            System.out.printf("  %-25s | Category: %s%n", name, category);
        }
    }

    /** Patient who can be prescribed multiple medications. */
    static class Patient {
        private String patientId;
        private String name;

        public Patient(String patientId, String name) {
            this.patientId = patientId;
            this.name = name;
        }

        public String getPatientId() { return patientId; }
        public String getName() { return name; }

        public void showInfo() {
            System.out.println("  Patient ID : " + patientId);
            System.out.println("  Name       : " + name);
        }
    }

    /** Represents a single prescription linking a patient to a medication. */
    static class Prescription {
        private Patient patient;
        private Medication medication;
        private LocalDate prescribedDate;

        public Prescription(Patient patient, Medication medication, LocalDate prescribedDate) {
            this.patient = patient;
            this.medication = medication;
            this.prescribedDate = prescribedDate;
        }

        public Patient getPatient() { return patient; }
        public Medication getMedication() { return medication; }

        public void showInfo() {
            System.out.printf("  %-12s | %-25s | Category: %-22s | Date: %s%n",
                    patient.getName(), medication.getName(), medication.getCategory(), prescribedDate);
        }
    }

    /** Forbidden medication category combinations. */
    static final String[][] FORBIDDEN_COMBINATIONS = {
        {"Antibiotics", "Statins"},
        {"Muscle Relaxants", "CNS Depressants"},
        {"Anti-Inflammatories", "Anticoagulants"},
    };

    /** Manages all prescriptions with conflict checking. */
    static class PrescriptionManager {
        private List<Prescription> prescriptions = new ArrayList<>();

        private Set<String> getPatientCategories(Patient patient) {
            Set<String> categories = new HashSet<>();
            for (Prescription p : prescriptions) {
                if (p.getPatient().getPatientId().equals(patient.getPatientId())) {
                    categories.add(p.getMedication().getCategory());
                }
            }
            return categories;
        }

        private String checkConflict(Set<String> existingCategories, String newCategory) {
            for (String[] pair : FORBIDDEN_COMBINATIONS) {
                if (newCategory.equals(pair[0]) && existingCategories.contains(pair[1])) {
                    return pair[0] + " cannot be combined with " + pair[1];
                }
                if (newCategory.equals(pair[1]) && existingCategories.contains(pair[0])) {
                    return pair[1] + " cannot be combined with " + pair[0];
                }
            }
            return "";
        }

        public void prescribe(Patient patient, Medication medication, LocalDate date) {
            // Check for duplicate
            for (Prescription p : prescriptions) {
                if (p.getPatient().getPatientId().equals(patient.getPatientId())
                        && p.getMedication().getName().equals(medication.getName())) {
                    System.out.println("  [Error] " + medication.getName() + " is already prescribed to " + patient.getName() + ".");
                    return;
                }
            }
            // Check for conflicts
            Set<String> existingCategories = getPatientCategories(patient);
            String conflict = checkConflict(existingCategories, medication.getCategory());
            if (!conflict.isEmpty()) {
                System.out.println("  [Error] Cannot prescribe " + medication.getName()
                        + " to " + patient.getName() + ": " + conflict + ".");
                return;
            }
            prescriptions.add(new Prescription(patient, medication, date));
            System.out.println("  [OK] " + medication.getName() + " prescribed to " + patient.getName() + ".");
        }

        public List<Prescription> getPatientPrescriptions(Patient patient) {
            List<Prescription> result = new ArrayList<>();
            for (Prescription p : prescriptions) {
                if (p.getPatient().getPatientId().equals(patient.getPatientId())) {
                    result.add(p);
                }
            }
            return result;
        }

        public List<Prescription> getMedicationPatients(Medication medication) {
            List<Prescription> result = new ArrayList<>();
            for (Prescription p : prescriptions) {
                if (p.getMedication().getName().equals(medication.getName())) {
                    result.add(p);
                }
            }
            return result;
        }

        public void showAll() {
            if (prescriptions.isEmpty()) {
                System.out.println("  No prescriptions recorded.");
                return;
            }
            System.out.printf("  %-12s | %-25s | %-22s | Date%n", "Patient", "Medication", "Category");
            System.out.println("  " + "-".repeat(80));
            for (Prescription p : prescriptions) {
                p.showInfo();
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        // --- Create medications ---
        Medication amoxicillin = new Medication("Amoxicillin", "Antibiotics");
        Medication ibuprofen = new Medication("Ibuprofen", "Anti-Inflammatories");
        Medication diazepam = new Medication("Diazepam", "Muscle Relaxants");
        Medication coughSyrup = new Medication("Dextromethorphan", "Cough Syrup");
        Medication insulin = new Medication("Insulin", "Insulin");
        Medication warfarin = new Medication("Warfarin", "Anticoagulants");
        Medication phenobarbital = new Medication("Phenobarbital", "CNS Depressants");
        Medication atorvastatin = new Medication("Atorvastatin", "Statins");

        System.out.println("=== Available Medications ===");
        Medication[] meds = {amoxicillin, ibuprofen, diazepam, coughSyrup,
                insulin, warfarin, phenobarbital, atorvastatin};
        for (Medication m : meds) {
            m.showInfo();
        }
        System.out.println();

        // --- Create patients ---
        Patient patient1 = new Patient("PAT-001", "Tareq");
        Patient patient2 = new Patient("PAT-002", "Afsana");
        Patient patient3 = new Patient("PAT-003", "Imtiaz");

        PrescriptionManager manager = new PrescriptionManager();
        LocalDate today = LocalDate.now();

        System.out.println("=== Prescribing Medications ===");
        manager.prescribe(patient1, amoxicillin, today);
        manager.prescribe(patient1, coughSyrup, today);
        manager.prescribe(patient1, atorvastatin, today); // CONFLICT
        System.out.println();

        manager.prescribe(patient2, diazepam, today);
        manager.prescribe(patient2, insulin, today);
        manager.prescribe(patient2, phenobarbital, today); // CONFLICT
        System.out.println();

        manager.prescribe(patient3, ibuprofen, today);
        manager.prescribe(patient3, coughSyrup, today);
        manager.prescribe(patient3, warfarin, today); // CONFLICT
        System.out.println();

        System.out.println("=== All Prescriptions ===");
        manager.showAll();

        Patient[] patients = {patient1, patient2, patient3};
        for (Patient p : patients) {
            System.out.println("=== Prescriptions for " + p.getName() + " ===");
            for (Prescription pr : manager.getPatientPrescriptions(p)) {
                pr.showInfo();
            }
            System.out.println();
        }

        System.out.println("=== Patients on Dextromethorphan (Cough Syrup) ===");
        for (Prescription p : manager.getMedicationPatients(coughSyrup)) {
            System.out.println("  " + p.getPatient().getName() + " (prescribed: " + p.prescribedDate + ")");
        }
        System.out.println();
    }
}
